src/misc/rwqueue.cpp

```diff
@@ -105,42 +142,58 @@ void RWQueue<T>::BulkEnqueue(std::vector<T>& from_source, const size_t num_reque
 		                                std::min(num_remaining,
 		                                         free_capacity));
 
-		// wait until the queue has enough room for the items
-		has_room.wait(lock, [&] { return capacity - queue.size() >= num_items; });
-
-		const auto source_end = source_start +
-		                        static_cast<difference_t>(num_items);
-		queue.insert(queue.end(),
-		             std::move_iterator(source_start),
-		             std::move_iterator(source_end));
+		// wait until we're stopped or the queue has enough room for the
+		// items
+		has_room.wait(lock, [&] {
+			return is_stopped || capacity - queue.size() >= num_items;
+		});
+
+		if (!is_stopped) {
+			const auto source_end = source_start +
+			                        static_cast<difference_t>(num_items);
+			queue.insert(queue.end(),
+			             std::move_iterator(source_start),
+			             std::move_iterator(source_end));
+			source_start = source_end;
+			num_remaining -= num_items;
+		} else {
+			// If we stopped while bulk enqueing, then stop here.
+			// Anything that was enqueued prior to being stopped is
+			// safely in the queue.
+			num_remaining = 0;
+		}
 
 		// notify the next waiting thread that we have an item
 		lock.unlock();
 		has_items.notify_one();
-
-		source_start = source_end;
-		num_remaining -= num_items;
 	}
 	from_source.clear();
+	return !is_stopped;
 }
 
 template <typename T>
 T RWQueue<T>::Dequeue()
 {
-	// wait until the queue has an item
+	// wait until we're stopped or the queue has an item
 	std::unique_lock<std::mutex> lock(mutex);
-	has_items.wait(lock, [this] { return !queue.empty(); });
-
-	// get it, and notify the first waiting thread that the queue has room
-	T item = std::move(queue.front());
-	queue.pop_front();
+	has_items.wait(lock, [this] { return is_stopped || !queue.empty(); });
+
+	T item = {};
+	// Even if the queue has stopped, we need to drain the (previously)
+	// queued items before we're done.
+	if (!is_stopped || !queue.empty()) {
```

Comment:
Subjectively, I'd use `is_running` because "not running" (=stopped) reads a lot better to me than "not stopped" (=running). The latter is almost like a double negative to me, and double negatives always give me a pause 😄 It's a very minor thing, though.

Reply:
`is_running` -> switched; thanks @johnnovak :+1: